src/otx/algo/diffusion/huggingface_model.py

```diff
@@ -0,0 +1,278 @@
+            raise ValueError(error_msg)
+
+        return {
+            "sample": noisy_latents,
+            "timestep": timesteps,
+            "encoder_hidden_states": encoder_hidden_states,
+        }
+
+    def _customize_outputs(
+        self,
+        outputs: torch.Tensor,
+        inputs: DiffusionBatchDataEntity,
+    ) -> DiffusionBatchPredEntity | OTXBatchLossEntity:
+        if torch.isnan(outputs).any():
+            error_msg = "NaN detected in the output."
+            raise ValueError(error_msg)
+        if not self.training:
+            msg = "This should never raise since `validation_step` is overridden."
+            raise NotImplementedError(msg)
+        return OTXBatchLossEntity(mse=F.mse_loss(outputs, self.target, reduction="mean"))
+
+    def on_fit_start(self) -> None:
+        """Called at the very beginning of fit.
+
+        If on DDP it is called on every process
+
+        """
+        self.pipe.to(self.device)
+        return super().on_fit_start()
+
+    def on_test_start(self) -> None:
+        """Called at the beginning of testing."""
+        self.pipe.to(self.device)
+        return super().on_test_start()
+
+    def on_predict_start(self) -> None:
+        """Called at the beginning of prediction."""
+        self.pipe.to(self.device)
+        return super().on_predict_start()
+
+    def validation_step(self, batch: DiffusionBatchDataEntity, batch_idx: int) -> torch.Tensor:
+        """Perform a single validation step on a batch of data from the validation set.
+
+        :param batch: A batch of data (a tuple) containing the input tensor of images and target
+            labels.
+        :param batch_idx: The index of the current batch.
+        """
+        # We use "latent" output type to overcome built-in NSFW filtering that blacks out some random images
+        images = self.pipe(batch.captions, output_type="latent").images
+        images = self.pipe.vae.decode(
+            images / self.pipe.vae.config.scaling_factor,
+            return_dict=False,
+        )[0]
+        images = self.pipe.image_processor.postprocess(
+            images,
+            output_type="pt",
+            do_denormalize=[True] * len(images),
+        )
+        self.metric.update(images, real=False)
+
+        # Save images
+        for logger in self.loggers:
+            if isinstance(logger, TensorBoardLogger):
+                for image, caption in zip(images, batch.captions):
+                    logger.experiment.add_image(f"val/images/{caption}", image.detach().cpu())
+        return images
+
+    def test_step(self, batch: DiffusionBatchDataEntity, batch_idx: int) -> None:
+        """Perform a single test step on a batch of data from the test set.
+
+        :param batch: A batch of data (a tuple) containing the input tensor of images and target
+            labels.
+        :param batch_idx: The index of the current batch.
+        """
+        return self.validation_step(batch, batch_idx)
```

Comment:
It seems similar with `OTXDiffusionModel`.

Reply:
Removed the `OTXDiffusionModel` duplicate.